include/bluetooth/hci_vs.h

```diff
@@ -300,6 +300,35 @@ struct bt_hci_evt_mesh_scanning_report {
 	struct bt_hci_evt_mesh_scan_report reports[0];
 } __packed;
 
+/* HCI Proprietary vendor events */
+#define BT_HCI_PROP_EVT_PREFIX                 0xFF
+
+struct bt_hci_evt_prop {
+	u8_t  prefix;
+} __packed;
```

Comment:
I couldn't find any place in the Core spec which would define the format of vendor specific events beyond the fact that 0xff is reserved as the vendor specific event code. Did I miss it, or you made the assumption based on how known vendor extensions are defined?

I'd rather not start forcing this kind of assumptions unless it's clearly stated in the core spec. Instead, I'm fine with adding a vendor specific event handler, which the app can register, and then the return code can indicate if the app handled the event or not. If it did not, then the stack can go through its own handlers if it finds a match. I.e. I wouldn't use the existing `struct event_handler` infrastructure for this. This is just based on some 10 minutes of looking at this PR and thinking about it, so I reserve the right to change my mind later :)

Reply:
I am far from an HCI expert here, and even less in the context of Zephyr, but glossing through the Core spec I can't find anything about this either. However, I do see this [hci_vs.h#L135](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/hci_vs.h#L135) which seem to indicate that there's already some Zephyr-specific "address space".

Reply:
I based this on how the mesh ext event in hci_vs.h was written (i know it is not used yet). Therefor i also just took the byte prefix 0xFF as i could see Mesh had taken 0xF0.

But i do think i like you'r proposal about just registering a handler for HCI vendor specific events, and then in the handler return success/failure in finding a proprietary subevent handler, then the hci_core can look at it's own handlers for any enabled HCI extension that the future brings.

Pseudo code-ish it would be along the lines of this:

```
app_vendor_specific_event_cb(buf) {
    check buf for proprietary subevent
    return 0 if handled else -EINVAL
}

hci_vendor_event(buf) {
    err = app_vendor_specific_event_cb(buf)
    if (err)
        default_vendor_specific_event_handler(buf)
}

```

Reply:
Event code 0xFF and OGF (opcode group field) 0x3F, i.e. all ones, are reserved for vendor specific debug commands and events. So i think using this should be fine, unless we intend to add debug commands/events as well. Should be safe as long as we know the controller is supporting our set of vendor specific commands/events.

Reply:
Just to make sure that we are aligned on how the HCI Prop Event is structured, i have tried to illustrate this:

```
Syntax: <byte offset>: <field> [<value>]

HCI Event (per core spec):
0: Event_Code
1: Parameter_Total_Length = N
2: Event_Parameter[N]


HCI Vendor-Specific Event (per core spec):
0: Event_Code = 0xFF
1: Parameter_Total_Length = N
2: Event_Parameter[N]


Zephyr HCI Vendor-Specific Event (per hci_vs.h):
0: Event_Code = 0xFF
1: Parameter_Total_Length = N
2: Subevent_Code
3: Subevent_Data[N-1]

Zephyr HCI Vendor-Specific Event: HCI Prop Report:
0: Event_Code = 0xFF
1: Parameter_Total_Length = N
2: Subevent_Code = 0xFF
3: Prop_Data_Length = N-2
4: Prop_Data[N-2]
```

Reply:
@thoh-ot it's true that the HCI Zephyr vendor spec follows a similar pattern as e.g. the LE Meta event, so for our stack-internal code we'll probably be able to take advantage of this. However if we're to introduce a public API to handle HCI vendor events it shouldn't IMO be build on this assumption. Unless I remember the meaning of the Kconfig options wrong, this also means that the vendor event handler registration shouldn't be dependent on `BT_HCI_VS` since that's specifically for *Zephyr* vendor extensions. Perhaps it'd make sense to rename it to `BT_HCI_VS_ZEPHYR` or something like that.

Reply:
@thoh-ot Thank you for that illustration.
The problem I see with this approach is that proprietary HCI events are still under Zephyr vendor specific extensions restrictions.
If a controller has implemented events under 0xFF, without any knowledge about about zephyr VS HCI extensions, then this handler cannot be used.
I think it would make more sense if they shared the SubEvent code ranges. That way you can use this without enabling Zephyr Vender Specific HCI Extensions. And controllers who have both, will have to know the sub-event ranges split.

Reply:
So what we need to do is document this in `hci.txt` first, in the documentation. The idea here is for multiple sets of vendor-specific command sets to coexist, including:

- Zephyr VS commands and events
- Mesh commands and events
- Vendor commands and events

In order to do that, the standard way of achieving it is to assign:

- An OCF for the whole set of commands of each category (like it's done in the mesh commands)
- An event prefix that is prepended to the actual event data.

We need to document this first, so that we know what we implement, and we make this compatible with the existing Mesh extensions.

The idea is the following, with some preliminary numbers:

Zephyr VS OGF: 0x3F
Zephyr VS OCF: 0x0000 - 0x0300
Zephyr VS Event Code: 0xFF
Zephyr VS Event Prefix: None, valid events are 0x00 - 0x80

Mesh VS OGF: 0x3F
Mesh VS OCF: 0x301
Mesh VS Event Code: 0xFF
Mesh VS Event prefix: 0x81

Vendor VS OGF; 0x3F
Vendor VS OCF: 0x302
Vendor VS Event Code: 0xFF
Vendor VS Event Prefix: 0x82

Let's discuss this further, but I would like this documented in hci.txt first.